### PR TITLE
Fix gRPC exception handling and improve traces

### PR DESCRIPTION
**Pull request by ggezer**
Client level retries were sometimes ineffective due to narrow handling of exceptions.

### CODE REVIEW

**Comment by ggezer** [comment]
jenkins, test this please

**Comment by ggezer** [comment]
jenkins, test this please

**Comment by ggezer** [comment]
please jenkins, test this please

**Comment by ggezer** [comment]
thank you jenkins, test this please

**Comment by ggezer** [comment]
jenkins, test this please

**Comment by ggezer** [comment]
@gpang This seems back to usual flakiness. I'll work more on some flakes, but I believe this is good to go for alleviating PR failures.

**Comment by ggezer** [comment]
jenkins, test this please

File: core/common/src/main/java/alluxio/grpc/GrpcChannel.java [hunk 1]
```diff
@@ -127,7 +127,7 @@ public boolean isHealthy() {
 
   @Override
   public String toString() {
-    return MoreObjects.toStringHelper(this)
+    return MoreObjects.toStringHelper("")
```
Comment (gpang) [inline core/common/src/main/java/alluxio/grpc/GrpcChannel.java]: why is this (and others) ""?

File: core/common/src/main/java/alluxio/grpc/GrpcChannel.java [hunk 2]
```diff
@@ -95,7 +95,7 @@ public void intercept(ClientInterceptor interceptor) {
   /**
    * Shuts down the channel.
    */
-  public void shutdown() {
+  public synchronized void shutdown() {
```
Comment (gpang) [inline core/common/src/main/java/alluxio/grpc/GrpcChannel.java]: could you add a comment in the code to explain why this has to be synchronized?

File: core/common/src/main/java/alluxio/security/authentication/AbstractSaslClientHandler.java [hunk 3]
```diff
@@ -85,10 +88,11 @@ public SaslMessage handleMessage(SaslMessage message) throws SaslException {
   }
 
   @Override
-  public void close() {
+  public synchronized void close() {
     if (mSaslClient != null) {
       try {
         mSaslClient.dispose();
+        mSaslClient = null;
```
Comment (gpang) [inline core/common/src/main/java/alluxio/security/authentication/AbstractSaslClientHandler.java]: do we want to keep this non-null if there is an issue with `dispose`? Basically, should we set it to null in finally or here?

File: core/common/src/main/java/alluxio/security/authentication/AbstractSaslClientHandler.java [hunk 4]
```diff
@@ -47,7 +47,10 @@ public AbstractSaslClientHandler(ChannelAuthenticationScheme authScheme) {
   }
 
   @Override
-  public SaslMessage handleMessage(SaslMessage message) throws SaslException {
+  public synchronized SaslMessage handleMessage(SaslMessage message) throws SaslException {
```
Comment (gpang) [inline core/common/src/main/java/alluxio/security/authentication/AbstractSaslClientHandler.java]: can you leave a comment as to why this and others have to be synchronized? (seems like it is for `mSaslClient`?)

File: core/common/src/main/java/alluxio/grpc/GrpcChannel.java (hunk 1 above)
Comment (ggezer) [reply core/common/src/main/java/alluxio/grpc/GrpcChannel.java]: When given an instance, the string generated will start with that class's name. (Something like "GrpcChannel { Host=....... }". Giving empty header makes the traces much cleaner.

File: core/common/src/main/java/alluxio/security/authentication/AbstractSaslClientHandler.java (hunk 3 above)
Comment (ggezer) [reply core/common/src/main/java/alluxio/security/authentication/AbstractSaslClientHandler.java]: We don't want to keep it set if dispose fails. So in finally is better.

File: core/common/src/main/java/alluxio/grpc/GrpcChannel.java (hunk 1 above)
Comment (gpang) [reply core/common/src/main/java/alluxio/grpc/GrpcChannel.java]: We want to remove the class name? I would have thought the class name is useful?

Comment (ggezer) [reply core/common/src/main/java/alluxio/grpc/GrpcChannel.java]: Yeah, I think it's better for this particular one. For the rest, empty is better.

File: core/common/src/main/java/alluxio/grpc/GrpcChannelKey.java [hunk 5]
```diff
@@ -288,7 +288,7 @@ public boolean equals(Object other) {
 
   @Override
   public String toString() {
-    return MoreObjects.toStringHelper(this)
+    return MoreObjects.toStringHelper("")
```
Comment (gpang) [inline core/common/src/main/java/alluxio/grpc/GrpcChannelKey.java]: I am confused when this is supposed to be "" or `this`?

File: core/common/src/main/java/alluxio/grpc/GrpcChannelKey.java [hunk 6]
```diff
@@ -305,7 +305,7 @@ public String toString() {
    * @return short representation of this channel key
    */
   public String toStringShort() {
-    return MoreObjects.toStringHelper("Channel")
+    return MoreObjects.toStringHelper("")
```
Comment (gpang) [inline core/common/src/main/java/alluxio/grpc/GrpcChannelKey.java]: I am confused when this is supposed to be "" or `this`?

File: core/common/src/main/java/alluxio/grpc/GrpcServer.java [hunk 7]
```diff
@@ -135,7 +135,7 @@ public boolean isServing() {
 
   @Override
   public String toString() {
-    return MoreObjects.toStringHelper(this)
+    return MoreObjects.toStringHelper("")
```
Comment (gpang) [inline core/common/src/main/java/alluxio/grpc/GrpcServer.java]: I am confused when this is supposed to be "" or `this`?

File: core/common/src/main/java/alluxio/grpc/GrpcServerAddress.java [hunk 8]
```diff
@@ -86,7 +86,7 @@ public boolean equals(Object other) {
 
   @Override
   public String toString() {
-    return MoreObjects.toStringHelper(this)
+    return MoreObjects.toStringHelper("")
```
Comment (gpang) [inline core/common/src/main/java/alluxio/grpc/GrpcServerAddress.java]: I am confused when this is supposed to be "" or `this`?

**Comment by ggezer** [comment]
jenkins, test this please

**Comment by ggezer** [comment]
alluxio-bot, test this please

**Comment by ggezer** [comment]
alluxio-bot, merge this please

**Pull request merged**